hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerSelectionCriteria.java

```diff
@@ -92,15 +96,31 @@ private boolean isContainerReplicatingOrDeleting(ContainerID containerID) {
    */
   public NavigableSet<ContainerID> getCandidateContainers(
       DatanodeDetails node, long sizeMovedAlready) {
-    NavigableSet<ContainerID> containerIDSet =
-        new TreeSet<>(orderContainersByUsedBytes().reversed());
+    // Initialize containerSet for node
+    if (!setMap.containsKey(node)) {
+      NavigableSet<ContainerID> newSet =
+          new TreeSet<>(orderContainersByUsedBytes().reversed());
+      try {
+        newSet.addAll(nodeManager.getContainers(node));
```

Comment:
IMO, sorting can be done on filtered containers only to avoid sorting of containers not meeting criteria.